File: javascript/robot-name/robot-name.js
const TOTAL_NAMES = 26 * 26 * 10 * 10 * 10
const LETTER1_DIV = 26 * 10 * 10 * 10
const LETTER2_DIV = 10 * 10 * 10
const NUM1_DIV = 10 * 10 
const NUM2_DIV = 10

const LETTER_A = 'A'.charCodeAt(0)

const robotLetter = (n, div) => String.fromCharCode(Math.floor(n / div) + LETTER_A)

const translateName = function (number) {
    const letter1 = robotLetter(number, LETTER1_DIV)
    let rem = number % LETTER1_DIV
    const letter2 = robotLetter(rem, LETTER2_DIV)
    rem = rem % LETTER2_DIV
    const num1 = Math.floor(rem / NUM1_DIV)
    rem = rem % NUM1_DIV
    const num2 = Math.floor(rem / NUM2_DIV)
    rem = rem % NUM2_DIV
    const num3 = rem

    return letter1 + letter2 + num1 + num2 + num3
}


export class Robot {
    constructor() {
        this._name = Robot.pickName()
    }

    reset() {
        this._name = Robot.pickName()
    }

    set name(v) {
        throw "internal name cannot be modified"
    }

    get name() {
        return this._name
    }
}

const initNames = () => {
    Robot.availableNames = []
    for (let i = 0; i < TOTAL_NAMES; i++) {
        Robot.availableNames.push(i)
    }
}
initNames();

Robot.pickName = () => {
    let idx = Math.floor(Math.random() * Math.floor(Robot.availableNames.length));
    let nameNumber = Robot.availableNames[idx]
    Robot.availableNames.splice(idx, 1)
    return translateName(nameNumber)
}


Robot.releaseNames = () => {
    initNames()
};

File: csharp/robot-name/RobotName.cs
using System;
using System.Collections.Generic;

public class Robot
{
    const int TOTAL_NAMES = 26 * 26 * 10 * 10 * 10;
    const int LETTER1_DIV = 26 * 10 * 10 * 10;
    const int LETTER2_DIV = 10 * 10 * 10;
    const int NUM1_DIV = 10 * 10;
    const int NUM2_DIV = 10;
    const char LETTER_A = 'A';
    static Random rnd = new Random();
    static List<int> availableNames = new List<int>();
    static Robot()
    {
        for (var i = 0; i < TOTAL_NAMES; i++)
        {
            Robot.availableNames.Add(i);
        }
    }

    static string RobotLetter(int n, int div)
    {
        char c = (char)(Math.Floor((double)n / div) + LETTER_A);
        return c.ToString();
    }

    static string TranslateName(int number)
    {
        string letter1 = RobotLetter(number, LETTER1_DIV);
        var rem = number % LETTER1_DIV;
        string letter2 = RobotLetter(rem, LETTER2_DIV);
        rem = rem % LETTER2_DIV;

        string num1 = Math.DivRem(rem, NUM1_DIV, out rem).ToString();
        string num2 = Math.DivRem(rem, NUM2_DIV, out rem).ToString();
        rem = rem % NUM2_DIV;
        string num3 = rem.ToString();

        return letter1 + letter2 + num1 + num2 + num3;
    }

    private string _name;
    public string Name
    {
        get
        {
            if (this._name != null)
            {
                return this._name;
            }

            var idx = rnd.Next(availableNames.Count);
            var nameNumber = Robot.availableNames[idx];
            Robot.availableNames.Remove(nameNumber);
            this._name = TranslateName(nameNumber);
            return this._name;

        }
    }

    public void Reset()
    {
        this._name = null;
    }
}